src/main/java/com/yahoo/bullet/storm/drpc/DRPCQuerySubscriber.java

```diff
@@ -0,0 +1,124 @@
+     * @param maxUnCommittedQueries The maximum number of queries that can be read without committing them.
+     */
+    public DRPCQuerySubscriber(BulletConfig config, int maxUnCommittedQueries) {
+        super(maxUnCommittedQueries);
+
+        collector = new DRPCOutputCollector();
+        emittedIDs = new HashMap<>();
+
+        // Get the Storm Config that has all the relevant cluster settings and properties
+        Map stormConfig = config.getRequiredConfigAs(DRPCConfig.STORM_CONFIG, Map.class);
+
+        // Get the TopologyContext
+        TopologyContext context = config.getRequiredConfigAs(DRPCConfig.STORM_CONTEXT, TopologyContext.class);
+
+        // Wrap the collector in a SpoutOutputCollector (it just delegates to the underlying DRPCOutputCollector)
+        SpoutOutputCollector spoutOutputCollector = new SpoutOutputCollector(collector);
+
+        // Get the DRPC function we should subscribe to
+        String function = config.getRequiredConfigAs(DRPCConfig.DRPC_FUNCTION, String.class);
+
+        spout = new DRPCSpout(function);
+        spout.open(stormConfig, context, spoutOutputCollector);
+    }
+
+    @Override
+    public List<PubSubMessage> getMessages() throws PubSubException {
+        // Try and read from DRPC. The DRPCSpout does a sleep for 1 ms if there are no tuples, so we don't have to do it.
+        spout.nextTuple();
+
+        if (!collector.haveOutput()) {
+            return null;
+        }
+
+        // The DRPCSpout only should have emitted one tuple
+        List<List<Object>> tuples = collector.reset();
+
+        log.debug("Have a message through DRPC {}", tuples);
+        List<Object> tupleAndID = tuples.get(0);
+
+        // The first object is the actual DRPCSpout tuple and the second is the DRPC messageID.
+        List<Object> tuple = (List<Object>) tupleAndID.get(0);
+        Object drpcID = tupleAndID.get(1);
+
+        // The first object in the tuple is our PubSubMessage as JSON
+        String pubSubMessageJSON = (String) tuple.get(0);
+        // The second object in the tuple is the serialized returnInfo added by the DRPCSpout
+        String returnInfo = (String) tuple.get(1);
+
+        log.debug("Read message\n{}\nfrom DRPC with return information {}", pubSubMessageJSON, returnInfo);
+
+        PubSubMessage pubSubMessage = PubSubMessage.fromJSON(pubSubMessageJSON);
+        // Add returnInfo as metadata. Cannot add it to pubSubMessage
+        String id = pubSubMessage.getId();
+        String content = pubSubMessage.getContent();
+        int sequence = pubSubMessage.getSequence();
+        PubSubMessage message = new PubSubMessage(id, content, new Metadata(null, returnInfo), sequence);
+
+        emittedIDs.put(ImmutablePair.of(id, sequence), drpcID);
+        return Collections.singletonList(message);
+    }
+
+    @Override
+    public void close() {
+        log.warn("Failing all pending requests: {}", emittedIDs);
+        emittedIDs.values().forEach(spout::fail);
```

Comment:
As we discussed, messages are not removed from emittedIDs when a message is ack'ed...  Override commit() and do that?

Reply:
Done